src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+            size_t keySize    = strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize);
+
+            VerifyOrExit(keySize < kMdnsTextMaxSize, error = CHIP_ERROR_INVALID_ARGUMENT);
+            entry.mKey = static_cast<char *>(chip::Platform::MemoryAlloc(keySize + 1));
+            VerifyOrExit(entry.mKey != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            strncpy(const_cast<char *>(entry.mKey), source.mTextEntryies[i].mKey, keySize + 1);
+            entry.mData = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(source.mTextEntryies[i].mDataSize));
+            VerifyOrExit(entry.mData != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            memcpy(const_cast<uint8_t *>(entry.mData), source.mTextEntryies[i].mData, entry.mDataSize);
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::Entry::Entry() : mHasValue(false), mHasPending(false), mNodeId(0), mFabricId(0)
+{
+    mService.mTextEntryies  = nullptr;
+    mService.mTextEntrySize = 0;
+}
+
+ServicePool::Entry::Entry(Entry && rhs)
+{
+    MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::operator=(Entry && rhs)
+{
+    return MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::MoveFrom(Entry && rhs)
+{
+    mService    = rhs.mService;
+    mNodeId     = rhs.mNodeId;
+    mFabricId   = rhs.mFabricId;
+    mHasValue   = rhs.mHasValue;
+    mHasPending = rhs.mHasPending;
+
+    rhs.mHasValue   = false;
+    rhs.mHasPending = false;
+
+    return *this;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(const MdnsService & service, uint64_t nodeId, uint64_t fabricId)
+{
+    CHIP_ERROR error;
+
+    Clear();
+    mService                          = service;
+    const_cast<uint64_t &>(mNodeId)   = nodeId;
+    const_cast<uint64_t &>(mFabricId) = fabricId;
+    mService.mTextEntryies            = nullptr;
+    error                             = CopyTextEntries(&mService, service);
+
+    if (error == CHIP_NO_ERROR)
+    {
+        mHasValue   = true;
+        mHasPending = false;
+    }
+    return error;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(MdnsService && service, uint64_t nodeId, uint64_t fabricId)
+{
+    Clear();
+    mService               = service;
+    mNodeId                = nodeId;
+    mFabricId              = fabricId;
+    service.mTextEntryies  = nullptr;
+    service.mTextEntrySize = 0;
+    mHasValue              = true;
+    mHasPending            = false;
+
+    return CHIP_NO_ERROR;
+}
+
+void ServicePool::Entry::Clear()
+{
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mHasValue   = false;
+    mHasPending = false;
+}
+
+ServicePool::Entry::~Entry()
+{
+    Clear();
```

Comment:
If we had a subclass for `CachedMdnsService`, its destructor could handle this, and we would not need this part at all.

Reply:
Tend to let the `Entry` handle all the memory since this will be the only place we need such things.

Reply:
My preference for having a `CachedMdnsService` is that then the type system can tell us which things have heap-allocated memory that we are responsible for....